pkg/server/service/loadbalancer/wrr/wrr.go

```diff
@@ -37,79 +36,63 @@ func New(sticky *dynamic.Sticky) *Balancer {
 	return balancer
 }
 
-// Balancer is a WeightedRoundRobin load balancer.
-type Balancer struct {
-	handlers []*namedHandler
-	mutex    *sync.Mutex
-	// Current index (starts from -1)
-	index         int
-	currentWeight int
-	stickyCookie  *stickyCookie
+// Len implements heap.Interface/sort.Interface
+func (b *Balancer) Len() int { return len(b.handlers) }
+
+// Less implements heap.Interface/sort.Interface
+func (b *Balancer) Less(i, j int) bool {
+	return b.handlers[i].deadline < b.handlers[j].deadline
 }
 
-func (b *Balancer) maxWeight() int {
-	max := -1
-	for _, s := range b.handlers {
-		if s.weight > max {
-			max = s.weight
-		}
-	}
-	return max
+// Swap implements heap.Interface/sort.Interface
+func (b *Balancer) Swap(i, j int) {
+	b.handlers[i], b.handlers[j] = b.handlers[j], b.handlers[i]
 }
 
-func (b *Balancer) weightGcd() int {
-	divisor := -1
-	for _, s := range b.handlers {
-		if divisor == -1 {
-			divisor = s.weight
-		} else {
-			divisor = gcd(divisor, s.weight)
-		}
+// Push implements heap.Interface for pushing an item into the heap
+func (b *Balancer) Push(x interface{}) {
+	h, ok := x.(*namedHandler)
+	if !ok {
+		return
 	}
-	return divisor
+	b.handlers = append(b.handlers, h)
 }
 
-func gcd(a, b int) int {
-	for b != 0 {
-		a, b = b, a%b
-	}
-	return a
+// Pop implements heap.Interface for poping an item from the heap
+func (b *Balancer) Pop() interface{} {
```

Comment:
Now that all the methods are on an exported type, it opens the door to panics due to misuses, such as calling b.Pop() on a freshly initialised Balancer. However, I don't think it is our job to make the methods robust to such misuses, since the callers have b.Len() that allows them to call b.Pop() (and others) responsibly.
However, let's at least document document that fact, by adding a line above, such as:

```
// It panics if b.Len() < 1.
```

I don't think we need to do it for Less and Swap, as the panics from them would be really obvious imho.